qutebrowser/misc/sql.py

```diff
@@ -168,75 +171,199 @@ def raise_sqlite_error(msg, error):
+#     QSqlDatabase.removeDatabase(QSqlDatabase.database().connectionName())
+#
+#
+# def version():
+#     """Return the sqlite version string."""
+#     try:
+#         if not QSqlDatabase.database().isOpen():
+#             init(':memory:')
+#             ver = Query("select sqlite_version()").run().value()
+#             close()
+#             return ver
+#         return Query("select sqlite_version()").run().value()
+#     except KnownError as e:
+#         return 'UNAVAILABLE ({})'.format(e)
+
+
+class Database:
+
+    """A wrapper over a QSqlDatabase connection."""
+
+    _USER_VERSION = UserVersion(0, 4)  # The current / newest user version
+
+    def __init__(self, path: str):
+        if QSqlDatabase.database(path).isValid():
+            raise BugError(f'A connection to the database at "{path}" already exists')
+
+        self._path = path
+        database = QSqlDatabase.addDatabase('QSQLITE', path)
+        if not database.isValid():
+            raise KnownError('Failed to add database. Are sqlite and Qt sqlite '
+                             'support installed?')
+        database.setDatabaseName(path)
+        if not database.open():
+            error = database.lastError()
+            msg = f"Failed to open sqlite database at {path}: {error.text()}"
+            raise_sqlite_error(msg, error)
+
+        version_int = self.query('pragma user_version').run().value()
+        self._user_version = UserVersion.from_int(version_int)
+
+        if self._user_version.major > self._USER_VERSION.major:
+            raise KnownError(
+                "Database is too new for this qutebrowser version (database version "
+                f"{self._user_version}, but {self._USER_VERSION.major}.x is supported)")
+
+        if self.user_version_changed():
+            # Enable write-ahead-logging and reduce disk write frequency
+            # see https://sqlite.org/pragma.html and issues #2930 and #3507
+            #
+            # We might already have done this (without a migration) in earlier versions,
+            # but as those are idempotent, let's make sure we run them once again.
+            self.query("PRAGMA journal_mode=WAL").run()
+            self.query("PRAGMA synchronous=NORMAL").run()
+
+    def qSqlDatabase(self):
+        database = QSqlDatabase.database(self._path, open=True)
+        if not database.isValid():
+            raise BugError('Failed to get connection. Did you close() this Database '
+                           'instance?')
+        return database
+
+    def query(self, querystr, forward_only=True):
+        return Query(self, querystr, forward_only)
+
+    def table(self, name, fields, constraints=None, parent=None):
+        return SqlTable(self, name, fields, constraints, parent)
+
+    def user_version_changed(self):
+        """Whether the version stored in the database differs from the current one."""
+        return self._user_version != self._USER_VERSION
+
+    def upgrade_user_version(self):
```

Comment:
Would this make more sense? A method that upgrades the user_version variable inside the database. The code using the database will call this **after** it has run the appropriate migrations.

I'm not sure that this improves things by a lot. If we currently crash during a migration, on the next run qutebrowser will think that everything went well and will operate over the old versions of the tables without knowing it, potentially crashing again (or losing data? It could be). With this approach qutebrowser will crash every time after a bad migration. The user won't be able to use it at all, but at least they can downgrade and the database will be untouched (I am running the migration code inside a transaction to avoid running a migration "partially"). 

If things start to get more complicated we *could* consider a full-fledged migration system. Something like Django's:
* Each migration is a separate SQL script.
* A migration is an operation at the database level, not at the table level. Our current approach is that each `SqlTable` subclass applies its own migrations when the user version changed, without them knowing about what the rest of the tables are doing.
* Each migration is run in a transaction, so the changes in table `History` would be rolled back if the changes in `CompletionHistory` fail.
* We have a `Migrations` table that records all the migrations run, in order, and let's us determine which are missing.

Reply:
> With this approach qutebrowser will crash every time after a bad migration.

Could you elaborate on that? I don't really follow, but perhaps there are two cases to consider:

1. qutebrowser crashes because of an unrelated issue (say, someone pulled the power plug during the migration, or the user got impatient somehow and killed qutebrowser)
   - Without those changes, qutebrowser runs over the old table without knowing it, resulting in crashes or potential data losses
   - With those changes, qutebrowser will run the migration again and hopefully succeed
2. qutebrowser crashes because the migration is buggy, or the data is somehow mis-formed to begin with
   - Without those changes, same as above
   - With those changes, the migration will probably fail again, and the user will give up (or hopefully ask for support)

I assume 1. is going to be much more common than 2., so this seems like a huge improvement?

Reply:
Ah, yes, I was mainly considering option 2. For option 1 you're right, this would be a much better behavior.

Reply:
Even for 2 I think it would be much better. I'll take a well defined failure over undefined behaviour (with potential data loss) any day.

Reply:
In that case we're on the same page =)